producer/s3.go

```diff
@@ -0,0 +1,349 @@
+}
+
+func (prod *S3) dropMessages(messages []core.Message) {
+	for _, msg := range messages {
+		prod.Drop(msg)
+	}
+}
+
+func (prod *S3) transformMessages(messages []core.Message) {
+	streamObjects := make(map[core.MessageStreamID]*objectData)
+
+	// Format and sort
+	for idx, msg := range messages {
+		msgData, streamID := prod.ProducerBase.Format(msg)
+
+		// Fetch buffer for this stream
+		objects, objectsExists := streamObjects[streamID]
+		if !objectsExists {
+			// Select the correct s3 path
+			s3Path, streamMapped := prod.streamMap[streamID]
+			if !streamMapped {
+				s3Path, streamMapped = prod.streamMap[core.WildcardStreamID]
+				if !streamMapped {
+					s3Path = core.StreamRegistry.GetStreamName(streamID)
+					prod.streamMap[streamID] = s3Path
+
+					shared.Metric.New(s3MetricMessages + s3Path)
+					shared.Metric.New(s3MetricMessagesSec + s3Path)
+					prod.counters[s3Path] = new(int64)
+				}
+			}
+
+			// split bucket from prefix in path
+			s3Bucket, s3Prefix := s3Path, ""
+			if strings.Contains(s3Path, "/") {
+				split := strings.SplitN(s3Path, "/", 2)
+				s3Bucket, s3Prefix = split[0], split[1]
+			}
+
+			// Create buffers for this s3 path
+			maxLength := len(messages) / prod.objectMaxMessages + 1
+			objects = &objectData{
+			    objects:            make([][]byte, 0, maxLength),
+				s3Bucket:           s3Bucket,
+				s3Path:				s3Path,
+				s3Prefix:           s3Prefix,
+				original:           make([][]*core.Message, 0, maxLength),
+				lastObjectMessages: 0,
+			}
+			streamObjects[streamID] = objects
+		}
+
+		// Fetch object for this buffer
+		objectExists := len(objects.objects) > 0
+		if !objectExists || objects.lastObjectMessages + 1 > prod.objectMaxMessages {
+			// Append object to stream
+			objects.objects = append(objects.objects, make([]byte, 0, len(msgData)))
+			objects.original = append(objects.original, make([]*core.Message, 0, prod.objectMaxMessages))
+			objects.lastObjectMessages = 0
+		} else {
+			objects.objects[len(objects.objects)-1] = append(objects.objects[len(objects.objects)-1], prod.delimiter...)
+		}
+
+		// Append message to object
+		objects.objects[len(objects.objects)-1] = append(objects.objects[len(objects.objects)-1], msgData...)
+		objects.lastObjectMessages += 1
+		objects.original[len(objects.original)-1] = append(objects.original[len(objects.original)-1], &messages[idx])
+	}
+
+	sleepDuration := prod.sendTimeLimit - time.Since(prod.lastSendTime)
+	if sleepDuration > 0 {
+		time.Sleep(sleepDuration)
+	}
+
+
+	// Send to S3
+	for _, objects := range streamObjects {
+		for idx, object := range objects.objects {
+			hash := sha1.Sum(object)
+			key := objects.s3Prefix + time.Now().Format(prod.timeWrite) + hex.EncodeToString(hash[:])
+			params := &s3.PutObjectInput{
+				Bucket:       aws.String(objects.s3Bucket),
+				Key:          aws.String(key),
+				Body:         bytes.NewReader(object),
+				StorageClass: aws.String(prod.storageClass),
```

Comment:
ContentType, ContentEncoding and Expires might also be useful fields here.
(But we could also add them later)